src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php

```diff
@@ -72,28 +75,31 @@ public function __construct(array $configuration, array $connectionCredentials =
         }
 
         $initializer = static function ($redis) use ($host, $port, $auth, $serializer, $dbIndex) {
-            $redis->connect($host, $port);
+            if (null === $redis) {
+                $redis = !\is_array($host) ? new \Redis() : new \RedisCluster(null, $host);
```

Comment:
why moving the instantiation inside the closure whcih forces everything to become nullable too ?

symfony/cache is defining separate `RedisProxy` and `RedisClusterProxy` btw. Maybe the same should be done for messenger.

Reply:
This is required because you can't use `connect()` on a `RedisCluster` instance. The connection credentials must be set [when you instanciate the class](https://github.com/phpredis/phpredis/blob/develop/cluster.markdown#declaring-a-cluster-with-an-array-of-seeds).

And I've just realized I've forgot to handle the `$auth` variable with the cluster 😅

Reply:
So, I've checked what `symfony/cache` does. Are you sure you want the same thing with `symfony/messenger`?

The way my code is currently written, I handle `Redis` and `RedisCluster` nearly the same way: both classes are instanciated in the initializer, things are pretty even without any surprises.

If you want me to add a `RedisClusterProxy`, then `Redis` and `RedisCluster` will not be handled the same way: the first one will be instanciated and then initialized afterwards, the second one will be instanciated and initialized at the same time. This seems a bit awkward.

Reply:
the thing is, `Redis` and `RedisCluster` have different capabilities. Keeping separate proxies allow to keep identifying which case we are dealing with in case we use such feature in the future (IIRC, that's the reason why symfony/cache did it like that)

Reply:
Well, I can do this if you want. However the initializer will still have to instanciate `RedisCluster`, the proxy will not be able to do so because it doesn't know the connection credentials and they must be provided on instanciation, thus the "initializer" will still be an "instanciator".

Or maybe you want separate initializers? One for `Redis` which acts just like the old one (before my PR) and another one for `RedisCluster` which returns an instance of the class?

Reply:
@stof When you have a bit of time, let me know what suites you, that way I can finish the PR for good. 😉

Reply:
I've just added a `RedisProxyCluster` class, let me know if there's any issue with it.